File: Workshop/csharp/EmployeeOnboarding/Onboarding.cs
using EmployeeOnboarding.Externals;
using EmployeeOnboarding.Models;

namespace EmployeeOnboarding;

public class Onboarding(IEmployeeRepository employees, IHrSystem hr, IItProvisioning it, IPayroll payroll)
{
    public OnboardingResult OnboardNewHire(AcceptedOffer offer)
    {
        try
        {
            var employee = employees.Register(offer);
            var contract = hr.GenerateContract(employee);
            var account = it.ProvisionAccount(contract);
            var enrollment = payroll.Enroll(account);
            return enrollment;
        }
        catch (EmployeeRegistrationException e)
        {
            throw new BusinessException(e.Message);
        }
        catch (ContractGenerationException e)
        {
            throw new BusinessException(e.Message);
        }
        catch (AccountProvisioningException e)
        {
            throw new BusinessException(e.Message);
        }
        catch (PayrollEnrollmentException e)
        {
            throw new BusinessException(e.Message);
        }
    }
}

File: Workshop/ts/src/onboarding.ts
import type {
  EmployeeRepository,
  HrSystem,
  ItProvisioning,
  Payroll,
} from "./externals/ports.js";
import type { AcceptedOffer, OnboardingResult } from "./models.js";
import {
  AccountProvisioningException,
  BusinessException,
  ContractGenerationException,
  EmployeeRegistrationException,
  PayrollEnrollmentException,
} from "./models.js";

export class Onboarding {
  constructor(
    private readonly employees: EmployeeRepository,
    private readonly hr: HrSystem,
    private readonly it: ItProvisioning,
    private readonly payroll: Payroll,
  ) {}

  onboardNewHire(offer: AcceptedOffer): OnboardingResult {
    try {
      const employee = this.employees.register(offer);
      const contract = this.hr.generateContract(employee);
      const account = this.it.provisionAccount(contract);
      const enrollment = this.payroll.enroll(account);
      return enrollment;
    } catch (e) {
      if (e instanceof EmployeeRegistrationException) {
        throw new BusinessException(e.message);
      }
      if (e instanceof ContractGenerationException) {
        throw new BusinessException(e.message);
      }
      if (e instanceof AccountProvisioningException) {
        throw new BusinessException(e.message);
      }
      if (e instanceof PayrollEnrollmentException) {
        throw new BusinessException(e.message);
      }
      throw e;
    }
  }
}
